File: autojs/src/main/java/com/stardust/autojs/execution/ExecutionConfig.java
package com.stardust.autojs.execution;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by Stardust on 2017/2/1.
 */
public class ExecutionConfig implements Serializable {

    public static final String TAG = "execution.config";
    private List<String> mRequirePath = Collections.emptyList();
    private String mExecutePath;
    public long delay = 0;
    public long interval = 0;
    public int loopTimes = 1;
    private Map<String, Object> mArguments = new HashMap<>();
    private int mIntentFlags = 0;

    public static ExecutionConfig getDefault() {
        return new ExecutionConfig();
    }

    public boolean runInNewThread = true;

    public ExecutionConfig runInNewThread(boolean runInNewThread) {
        this.runInNewThread = runInNewThread;
        return this;
    }

    public int getIntentFlags() {
        return mIntentFlags;
    }

    public ExecutionConfig setIntentFlags(int intentFlags) {
        mIntentFlags = intentFlags;
        return this;
    }

    public ExecutionConfig requirePath(String... requirePath) {
        mRequirePath = new ArrayList<>(Arrays.asList(requirePath));
        if (mExecutePath != null) {
            mRequirePath.add(mExecutePath);
        }
        return this;
    }

    public ExecutionConfig executePath(String executePath) {
        mExecutePath = executePath;
        return this;
    }

    public List<String> getRequirePath() {
        return mRequirePath;
    }

    public String getExecutePath() {
        return mExecutePath;
    }

    public void setArgument(String key, Object object){
        mArguments.put(key, object);
    }

    public Object getArgument(String key){
        return mArguments.get(key);
    }

    public Map<String, Object> getArguments() {
        return mArguments;
    }

    public ExecutionConfig loop(long delay, int loopTimes, long interval) {
        this.delay = delay;
        this.loopTimes = loopTimes;
        this.interval = interval;
        return this;
    }
}

File: autojs/src/main/java/com/stardust/autojs/execution/ExecutionConfig.kt
package com.stardust.autojs.execution

import android.os.Parcel
import android.os.Parcelable
import com.stardust.autojs.project.ScriptConfig
import java.util.*

/**
 * Created by Stardust on 2017/2/1.
 */
data class ExecutionConfig(var workingDirectory: String = "",
                           var path: Array<out String> = emptyArray(),
                           var intentFlags: Int = 0,
                           var delay: Long = 0,
                           var interval: Long = 0,
                           var loopTimes: Int = 1,
                           var scriptConfig: ScriptConfig = ScriptConfig()) : Parcelable {


    private val mArguments = HashMap<String, Any>()

    val arguments: Map<String, Any>
        get() = mArguments

    constructor(parcel: Parcel) : this(
            parcel.readString().orEmpty(),
            parcel.createStringArray().orEmpty(),
            parcel.readInt(),
            parcel.readLong(),
            parcel.readLong(),
            parcel.readInt())

    fun setArgument(key: String, `object`: Any) {
        mArguments[key] = `object`
    }

    fun getArgument(key: String): Any? {
        return mArguments[key]
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ExecutionConfig

        if (workingDirectory != other.workingDirectory) return false
        if (!path.contentEquals(other.path)) return false
        if (intentFlags != other.intentFlags) return false
        if (delay != other.delay) return false
        if (interval != other.interval) return false
        if (loopTimes != other.loopTimes) return false
        if (mArguments != other.mArguments) return false

        return true
    }

    override fun hashCode(): Int {
        var result = workingDirectory.hashCode()
        result = 31 * result + path.contentHashCode()
        result = 31 * result + intentFlags
        result = 31 * result + delay.hashCode()
        result = 31 * result + interval.hashCode()
        result = 31 * result + loopTimes
        result = 31 * result + mArguments.hashCode()
        return result
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(workingDirectory)
        parcel.writeStringArray(path)
        parcel.writeInt(intentFlags)
        parcel.writeLong(delay)
        parcel.writeLong(interval)
        parcel.writeInt(loopTimes)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<ExecutionConfig> {

        @JvmStatic
        val tag = "execution.config"

        @JvmStatic
        val default: ExecutionConfig
            get() = ExecutionConfig()

        override fun createFromParcel(parcel: Parcel): ExecutionConfig {
            return ExecutionConfig(parcel)
        }

        override fun newArray(size: Int): Array<ExecutionConfig?> {
            return arrayOfNulls(size)
        }
    }
}
